### DISCUSSION TITLE
Linker Problems when Trying to Use C API

### DISCUSSION THREAD

**Discussion by sven1904** [bulletphysics/bullet3#3895] [Q&A] [answered]
Hello,

i would like to build an application that makes use of the Bullet C API.
Unfortunately, I always get linker errors although i link the bullet libraries.

Here's my minimum (not) working example:

CMakeLists.txt:
```
cmake_minimum_required(VERSION 3.4)

project(MWE)

find_package(Bullet REQUIRED)

include_directories(${BULLET_INCLUDE_DIRS})

add_executable(bullet-connect bullet-connect.cpp)
target_link_libraries(bullet-connect ${BULLET_LIBRARIES}
                                     # BulletRobotics
)
```

bullet-connect.cpp:
```
#include <bullet/SharedMemory/PhysicsClientC_API.h>

int main(int argc, char** argv)
{
  b3ConnectPhysicsDirect();

  return 0;
}
```

When I try to build it with `cmake . && make`, it gives me the following error:
```
/usr/bin/ld: CMakeFiles/bullet-connect.dir/bullet-connect.cpp.o: in function `main':
bullet-connect.cpp:(.text+0x14): undefined reference to `b3ConnectPhysicsDirect'
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/bullet-connect.dir/build.make:88: bullet-connect] Error 1
make[2]: *** [CMakeFiles/Makefile2:76: CMakeFiles/bullet-connect.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/bullet-connect.dir/rule] Error 2
make: *** [Makefile:118: bullet-connect] Error 2
```

Uncommenting `BulletRobotics` in CMakeLists.txt resolves that undefined reference issue for `b3ConnectPhysicsDirect` but causes hundreds of other linker errors, which i could not resolve yet:
[log.txt](https://github.com/bulletphysics/bullet3/files/6664160/log.txt)

Please, could you guide me in the right direction how to use the C API correctly?

Further information:
- bullet has been build with and installed with `cmake . && make && make install`
- OS: Ubuntu 20.04

**Comment by erwincoumans** [comment]
You need more dependencies, dependent on use of GUI/graphics (GUI) or pure headless (SHARED_MEMORY, DIRECT)

For DIRECT, TARGET_LINK_LIBRARIES(App_RobotSimulator_NoGUI BulletRobotics BulletFileLoader BulletWorldImporter BulletSoftBody BulletDynamics BulletCollision BulletInverseDynamicsUtils BulletInverseDynamics LinearMath Bullet3Common)

https://github.com/bulletphysics/bullet3/blob/master/examples/RobotSimulator/CMakeLists.txt

**Reply by sven1904** [reply]
Thank you for the reply.

Some of the errors were resolved with the additional libraries as well as `dl` and `pthread`.
Unfortunately, there are still some errors left: [log.txt](https://github.com/bulletphysics/bullet3/files/6747609/log.txt)
And all of them seem to be related to soft body collisions.

Especially, what i don't get is that there are errors related to the files in my local bullet3 repository although i did `make install` for bullet and reference the installed libraries and include directories in my `CMakeLists.txt`.
And when i change the path of the local repository, the errors are exactly the same, although these files in which the errors are reported do not even exist anymore.

I would appreciate any help.

In the end, I would also like to use a GUI since I like the possibilities of mouse interaction in the bullet examples, but for the beginning, it would be sufficient to get the simple non-GUI code posted in the question to compile and run.

**Comment by sven1904** [comment, marked as answer]
Finally, I could solve the linker errors.

They were caused by the wrong order of the library entries in `${BULLET_LIBRARIES}`.

The content of the variable looked like this:
`/usr/local/lib/libBulletDynamics.a;/usr/local/lib/libBulletCollision.a;/usr/local/lib/libLinearMath.a;/usr/local/lib/libBulletSoftBody.a`

But the (at least for me) correct order is this:
`/usr/local/lib/libBulletDynamics.a;/usr/local/lib/libBulletSoftBody.a;/usr/local/lib/libBulletCollision.a;/usr/local/lib/libLinearMath.a`

This is the same order as in @erwincoumans' answer, but, unfortunately, his suggestion did not work for me since i could not use the pure library name.
I really had to use `/usr/local/lib/libSomething.a`, which is strange since the other libraries (Bullet3Common, BulletRobotics, etc.) could be used with only their names.

Anyway, thanks again for your help @erwincoumans.
And, most probably, i will come back to the discussion pages when I enter the GUI universe :sweat_smile:.
